tfjs-backend-webgpu/src/kernels/SparseToDense.ts

```diff
@@ -46,24 +49,91 @@ export function sparseToDense(args: {
         sliceRank, strides, $defaultValue, sumDupeIndices);
     return backend.makeTensorInfo(outputShape, outBuf.dtype, outBuf.values);
   }
+
+  // If |sparseValues| is a scalar tensor, broadcast to the appropriate shape.
+  const inferredSparseValuesShape = sparseIndices.shape.length > 1 ?
+      sparseIndices.shape.slice(0, -1).concat(outputShape.slice(sliceRank)) :
+      sparseIndices.shape.slice();
+  const dimensionalSparseValues =
+      sparseValues.shape.length === inferredSparseValuesShape.length ?
+      identity({inputs: {x: sparseValues}, backend}) :
+      reshape({
+        inputs: {x: sparseValues},
+        backend,
+        attrs: {shape: new Array(inferredSparseValuesShape.length).fill(1)}
+      });
+  const $sparseValues =
+      util.arraysEqual(
+          inferredSparseValuesShape, dimensionalSparseValues.shape) ?
+      identity({inputs: {x: dimensionalSparseValues}, backend}) :
+      tile({
+        inputs: {x: dimensionalSparseValues},
+        backend,
+        attrs: {reps: inferredSparseValuesShape}
+      });
+  backend.disposeData(dimensionalSparseValues.dataId);
+
+  const flattenShape = [outputSize / sliceSize, sliceSize];
+
+  const flattenIndices = reshape({
+    inputs: {x: sparseIndices},
+    backend,
+    attrs: {shape: [numUpdates, sliceRank]}
+  });
+  const flattenValues = reshape({
+    inputs: {x: $sparseValues},
+    backend,
+    attrs: {shape: [numUpdates, sliceSize]}
+  });
+
+  const type = flattenValues.dtype;
+
+  // Fill output tensor with the default value.
+  const dimensionalDefaultValue = reshape({
+    inputs: {x: defaultValue},
+    backend,
+    attrs: {shape: new Array(flattenShape.length).fill(1)}
+  });
+  const flattenOutput = tile({
+    inputs: {x: dimensionalDefaultValue},
+    backend,
+    attrs: {reps: flattenShape}
+  });
+  backend.disposeData(dimensionalDefaultValue.dataId);
+
+  const size = util.sizeFromShape(flattenValues.shape);
   const uniformData = [
-    {type: 'int32', data: [numUpdates]},
     {type: 'int32', data: [sliceRank]},
     {type: 'int32', data: strides},
+    {type: 'int32', data: [size]},
   ];
-  const program = new ScatterProgram(
-      numUpdates, sliceRank, sparseIndices.shape.length,
-      sparseValues.shape.length, strides, [outputSize, 1], sumDupeIndices);
+  const program = new ScatterOptimizedProgram(
+      flattenValues.shape, sliceRank, flattenIndices.shape.length,
+      flattenValues.shape.length, strides, flattenShape, type);
+
+  // First replace the default value with 0 at indices.
```

Comment:
Is it better to write a new program (fillZeroAtIndices?) to process this pass to fill zero at indices?
Two advantages: 1) Don't need to add extra complexity for original `ScatterOptimizedProgram`. 2) For this pass itself, it may have better performance.

Reply:
I'm not sure. Probably the main tradeoff is

```
shader compilation * 1 + unoptimized shader execution * N
```

vs

```
shader compilation * 2 + optimized shader execution * N
```

Reply:
I vote this one.
> 
> ```
> shader compilation * 2 + optimized shader execution * N
> ```